commit 245bab17e614fd1e6c1be8d6fc96f937677de847
Author: askmeaboutloom <askmeaboutlo0m@users.noreply.github.com>
Date:   Thu Aug 13 15:23:22 2026 +0200

    [android] Handle edge-to-edge by not doing it
    
    Under Android 16 (API level 36, Baklava), edge-to-edge display is
    enforced. That means Android will always spill the application view into
    nonsensical areas like underneath the notification bar, the navigation
    bar or the camera notch, with no OS-level way to opt out of it.
    
    This commit deals with it by returning the behavior to how it was
    before, adding blank borders around the areas where those elements are.
    This only gets applied on Android 16, so the application must still
    define android:windowOptOutEdgeToEdgeEnforcement so that this works
    properly on Android 15! Earlier Android versions don't enable
    edge-to-edge display by default, so no need to do anything there.
    
    Implementation-wise, this installs a listener for "insets" on the root
    layout and passes the dimensions into Qt so that it can use them for the
    platform screen's available geometry. The existing code played pretty
    loose with the geometry and available geometry since they were always
    the same, so some things had to be shifted over from one to the other.

--- a/src/android/jar/src/org/qtproject/qt5/android/QtActivityDelegate.java
+++ b/src/android/jar/src/org/qtproject/qt5/android/QtActivityDelegate.java
@@ -187,6 +187,7 @@ public class QtActivityDelegate
             m_activity.getWindow().clearFlags(WindowManager.LayoutParams.FLAG_FULLSCREEN);
             m_activity.getWindow().getDecorView().setSystemUiVisibility(View.SYSTEM_UI_FLAG_VISIBLE);
         }
+        m_layout.setFullScreen(m_fullScreen);
         m_layout.requestLayout();
     }
 
diff --git a/src/android/jar/src/org/qtproject/qt5/android/QtLayout.java b/src/android/jar/src/org/qtproject/qt5/android/QtLayout.java
index 8b71cf353a..a90ce4090e 100644
--- a/src/android/jar/src/org/qtproject/qt5/android/QtLayout.java
+++ b/src/android/jar/src/org/qtproject/qt5/android/QtLayout.java
@@ -45,24 +45,33 @@ import android.content.Context;
 import android.os.Build;
 import android.util.AttributeSet;
 import android.util.DisplayMetrics;
+import android.util.Log;
 import android.view.View;
 import android.view.ViewGroup;
+import android.graphics.Insets;
 import android.graphics.Rect;
 import android.view.InputDevice;
 import android.view.MotionEvent;
 import android.view.GestureDetector;
+import android.view.WindowInsets;
 
 public class QtLayout extends ViewGroup
 {
     private GestureDetector m_gestureDetector;
     private Runnable m_startApplicationRunnable;
     private int m_bottomDisplayFrame = -1;
+    private int m_insetLeft = 0;
+    private int m_insetTop = 0;
+    private int m_insetRight = 0;
+    private int m_insetBottom = 0;
+    private boolean m_fullScreen = false;
     private QtInputEventDispatcher inputEventDispatcher;
 
     public QtLayout(Context context, Runnable startRunnable)
     {
         super(context);
         m_startApplicationRunnable = startRunnable;
+        initializeInsets();
         initializeFocus();
         initializeGestureDetector(context);
         inputEventDispatcher = QtNative.getInputEventDispatcher();
@@ -71,6 +80,7 @@ public class QtLayout extends ViewGroup
     public QtLayout(Context context, AttributeSet attrs)
     {
         super(context, attrs);
+        initializeInsets();
         initializeFocus();
         initializeGestureDetector(context);
         inputEventDispatcher = QtNative.getInputEventDispatcher();
@@ -79,11 +89,46 @@ public class QtLayout extends ViewGroup
     public QtLayout(Context context, AttributeSet attrs, int defStyle)
     {
         super(context, attrs, defStyle);
+        initializeInsets();
         initializeFocus();
         initializeGestureDetector(context);
         inputEventDispatcher = QtNative.getInputEventDispatcher();
     }
 
+    public void setFullScreen(boolean fullScreen)
+    {
+        m_fullScreen = fullScreen;
+    }
+
+    private void initializeInsets()
+    {
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.BAKLAVA) {
+            try {
+                setOnApplyWindowInsetsListener(new View.OnApplyWindowInsetsListener() {
+                    @Override
+                    public WindowInsets onApplyWindowInsets(View v, WindowInsets windowInsets) {
+                        try {
+                            int typeMask = WindowInsets.Type.displayCutout();
+                            if (!m_fullScreen) {
+                                typeMask |= WindowInsets.Type.systemBars();
+                            }
+                            Insets insets = windowInsets.getInsets(typeMask);
+                            m_insetLeft = insets.left;
+                            m_insetTop = insets.top;
+                            m_insetRight = insets.right;
+                            m_insetBottom = insets.bottom;
+                        } catch (Exception e) {
+                            Log.e(QtNative.QtTAG, "DPINPUT Error applying insets", e);
+                        }
+                        return windowInsets;
+                    }
+                });
+            } catch (Exception e) {
+                Log.e(QtNative.QtTAG, "DPINPUT Error setting insets listener", e);
+            }
+        }
+    }
+
     private void initializeFocus()
     {
         setFocusable(true);
@@ -121,7 +166,11 @@ public class QtLayout extends ViewGroup
 
         if (m_bottomDisplayFrame !=  r.bottom || oldh == -1) {
             m_bottomDisplayFrame =  r.bottom;
-            QtNative.setApplicationDisplayMetrics(metrics.widthPixels, metrics.heightPixels, w, h,
+            QtNative.setApplicationDisplayMetrics(metrics.widthPixels, metrics.heightPixels,
+                                                m_insetLeft,
+                                                m_insetTop,
+                                                w - m_insetLeft - m_insetRight,
+                                                h - m_insetTop - m_insetBottom,
                                                 metrics.xdpi,
                                                 metrics.ydpi,
                                                 metrics.scaledDensity,
--- a/src/android/jar/src/org/qtproject/qt5/android/QtNative.java
+++ b/src/android/jar/src/org/qtproject/qt5/android/QtNative.java
@@ -104,6 +104,8 @@ public class QtNative
     private static boolean m_isKeyboardHiding = false;
     private static int m_displayMetricsScreenWidthPixels = 0;
     private static int m_displayMetricsScreenHeightPixels = 0;
+    private static int m_displayMetricsDesktopXPixels = 0;
+    private static int m_displayMetricsDesktopYPixels = 0;
     private static int m_displayMetricsDesktopWidthPixels = 0;
     private static int m_displayMetricsDesktopHeightPixels = 0;
     private static double m_displayMetricsXDpi = .0;
@@ -369,6 +371,8 @@ public class QtNative
                     res[0] = startQtAndroidPlugin(qtParams, environment);
                     setDisplayMetrics(m_displayMetricsScreenWidthPixels,
                                       m_displayMetricsScreenHeightPixels,
+                                      m_displayMetricsDesktopXPixels,
+                                      m_displayMetricsDesktopYPixels,
                                       m_displayMetricsDesktopWidthPixels,
                                       m_displayMetricsDesktopHeightPixels,
                                       m_displayMetricsXDpi,
@@ -392,6 +396,8 @@ public class QtNative
 
     public static void setApplicationDisplayMetrics(int screenWidthPixels,
                                                     int screenHeightPixels,
+                                                    int desktopXPixels,
+                                                    int desktopYPixels,
                                                     int desktopWidthPixels,
                                                     int desktopHeightPixels,
                                                     double XDpi,
@@ -410,6 +416,8 @@ public class QtNative
             if (m_started) {
                 setDisplayMetrics(screenWidthPixels,
                                   screenHeightPixels,
+                                  desktopXPixels,
+                                  desktopYPixels,
                                   desktopWidthPixels,
                                   desktopHeightPixels,
                                   XDpi,
@@ -420,6 +428,8 @@ public class QtNative
             } else {
                 m_displayMetricsScreenWidthPixels = screenWidthPixels;
                 m_displayMetricsScreenHeightPixels = screenHeightPixels;
+                m_displayMetricsDesktopXPixels = desktopXPixels;
+                m_displayMetricsDesktopYPixels = desktopYPixels;
                 m_displayMetricsDesktopWidthPixels = desktopWidthPixels;
                 m_displayMetricsDesktopHeightPixels = desktopHeightPixels;
                 m_displayMetricsXDpi = XDpi;
@@ -997,6 +1007,8 @@ public class QtNative
     // screen methods
     public static native void setDisplayMetrics(int screenWidthPixels,
                                                 int screenHeightPixels,
+                                                int desktopXPixels,
+                                                int desktopYPixels,
                                                 int desktopWidthPixels,
                                                 int desktopHeightPixels,
                                                 double XDpi,
--- a/src/android/jar/src/org/qtproject/qt5/android/QtServiceDelegate.java
+++ b/src/android/jar/src/org/qtproject/qt5/android/QtServiceDelegate.java
@@ -115,7 +115,7 @@ public class QtServiceDelegate
         QtNative.setService(m_service, this);
         QtNative.setClassLoader(classLoader);
 
-        QtNative.setApplicationDisplayMetrics(10, 10, 10, 10, 120, 120, 1.0, 1.0, false);
+        QtNative.setApplicationDisplayMetrics(10, 10, 0, 0, 10, 10, 120, 120, 1.0, 1.0, false);
 
         if (loaderParams.containsKey(STATIC_INIT_CLASSES_KEY)) {
             for (String className: loaderParams.getStringArray(STATIC_INIT_CLASSES_KEY)) {
diff --git a/src/plugins/platforms/android/androidjniinput.cpp b/src/plugins/platforms/android/androidjniinput.cpp
index 294e98c03c..85a7183da3 100644
--- a/src/plugins/platforms/android/androidjniinput.cpp
+++ b/src/plugins/platforms/android/androidjniinput.cpp
@@ -366,13 +366,15 @@ namespace QtAndroidInput
     static void touchAdd(int id, Qt::TouchPointState state, bool primary, int x, int y,
         float major, float minor, float rotation, float pressure)
     {
+        const int dx = desktopXPixels();
+        const int dy = desktopYPixels();
         const int dw = desktopWidthPixels();
         const int dh = desktopHeightPixels();
         QWindowSystemInterface::TouchPoint touchPoint;
         touchPoint.id = id;
         touchPoint.pressure = pressure;
         touchPoint.rotation = qRadiansToDegrees(rotation);
-        touchPoint.normalPosition = QPointF(double(x) / dw, double(y) / dh);
+        touchPoint.normalPosition = QPointF(double(x - dx) / dw, double(y - dy) / dh);
         touchPoint.state = state;
 
         touchPoint.area = QRectF(x - double(minor),
@@ -560,6 +562,8 @@ namespace QtAndroidInput
                        && pointerIndex == actionIndex)) {
             return Qt::TouchPointReleased;
         } else {
+            const int dx = desktopXPixels();
+            const int dy = desktopYPixels();
             const int dw = desktopWidthPixels();
             const int dh = desktopHeightPixels();
 
@@ -568,8 +572,8 @@ namespace QtAndroidInput
             if (m_lastTouchPointsPos.contains(id)) {
                 lastNormalPosition = m_lastTouchPointsPos[id];
             }
-            const QPointF currentNormalPosition = QPointF(double(event.getX(pointerIndex)) / dw,
-                                                          double(event.getX(pointerIndex)) / dh);
+            const QPointF currentNormalPosition = QPointF(double(event.getX(pointerIndex) - dx) / dw,
+                                                          double(event.getY(pointerIndex) - dy) / dh);
             m_lastTouchPointsPos[id] = currentNormalPosition;
 
             return (currentNormalPosition == lastNormalPosition) ? Qt::TouchPointStationary
--- a/src/plugins/platforms/android/androidjnimain.cpp
+++ b/src/plugins/platforms/android/androidjnimain.cpp
@@ -113,6 +113,8 @@ static int m_surfaceId = 1;
 
 static QAndroidPlatformIntegration *m_androidPlatformIntegration = nullptr;
 
+static int m_desktopXPixels = 0;
+static int m_desktopYPixels = 0;
 static int m_desktopWidthPixels  = 0;
 static int m_desktopHeightPixels = 0;
 static double m_scaledDensity = 0;
@@ -162,6 +164,16 @@ namespace QtAndroid
                : 0;
     }
 
+    int desktopXPixels()
+    {
+        return m_desktopXPixels;
+    }
+
+    int desktopYPixels()
+    {
+        return m_desktopYPixels;
+    }
+
     int desktopWidthPixels()
     {
         return m_desktopWidthPixels;
@@ -659,6 +671,7 @@ static void setSurface(JNIEnv *env, jobject /*thiz*/, jint id, jobject jSurface,
 
 static void setDisplayMetrics(JNIEnv */*env*/, jclass /*clazz*/,
                             jint widthPixels, jint heightPixels,
+                            jint desktopXPixels, jint desktopYPixels,
                             jint desktopWidthPixels, jint desktopHeightPixels,
                             jdouble xdpi, jdouble ydpi,
                             jdouble scaledDensity, jdouble density, bool forceUpdate)
@@ -666,10 +679,14 @@ static void setDisplayMetrics(JNIEnv */*env*/, jclass /*clazz*/,
     // Android does not give us the correct screen size for immersive mode, but
     // the surface does have the right size
 
-    bool updateDesktopSize = m_desktopWidthPixels != desktopWidthPixels;
+    bool updateDesktopSize = m_desktopXPixels != desktopXPixels
+            || m_desktopYPixels != desktopYPixels || m_desktopWidthPixels != desktopWidthPixels
+            || m_desktopHeightPixels != desktopHeightPixels;
     widthPixels = qMax(widthPixels, desktopWidthPixels);
     heightPixels = qMax(heightPixels, desktopHeightPixels);
 
+    m_desktopXPixels = desktopXPixels;
+    m_desktopYPixels = desktopYPixels;
     m_desktopWidthPixels = desktopWidthPixels;
     m_desktopHeightPixels = desktopHeightPixels;
     m_scaledDensity = scaledDensity;
@@ -688,7 +705,8 @@ static void setDisplayMetrics(JNIEnv */*env*/, jclass /*clazz*/,
                                                         qRound(double(heightPixels) / ydpi * 25.4));
         m_androidPlatformIntegration->setScreenSize(widthPixels, heightPixels);
         if (updateDesktopSize || forceUpdate) {
-            m_androidPlatformIntegration->setDesktopSize(desktopWidthPixels, desktopHeightPixels);
+            m_androidPlatformIntegration->setDesktopSize(desktopXPixels, desktopYPixels,
+                                                         desktopWidthPixels, desktopHeightPixels);
         }
     }
 }
@@ -815,7 +833,7 @@ static JNINativeMethod methods[] = {
     {"quitQtCoreApplication", "()V", (void *)quitQtCoreApplication},
     {"terminateQt", "()V", (void *)terminateQt},
     {"waitForServiceSetup", "()V", (void *)waitForServiceSetup},
-    {"setDisplayMetrics", "(IIIIDDDDZ)V", (void *)setDisplayMetrics},
+    {"setDisplayMetrics", "(IIIIIIDDDDZ)V", (void *)setDisplayMetrics},
     {"setSurface", "(ILjava/lang/Object;II)V", (void *)setSurface},
     {"updateWindow", "()V", (void *)updateWindow},
     {"updateApplicationState", "(I)V", (void *)updateApplicationState},
--- a/src/plugins/platforms/android/androidjnimain.h
+++ b/src/plugins/platforms/android/androidjnimain.h
@@ -77,6 +77,8 @@ namespace QtAndroid
     void bringChildToBack(int surfaceId);
 
     QWindow *topLevelWindowAt(const QPoint &globalPos);
+    int desktopXPixels();
+    int desktopYPixels();
     int desktopWidthPixels();
     int desktopHeightPixels();
     double scaledDensity();
--- a/src/plugins/platforms/android/qandroidplatformintegration.cpp
+++ b/src/plugins/platforms/android/qandroidplatformintegration.cpp
@@ -476,10 +476,10 @@ QPlatformAccessibility *QAndroidPlatformIntegration::accessibility() const
 }
 #endif
 
-void QAndroidPlatformIntegration::setDesktopSize(int width, int height)
+void QAndroidPlatformIntegration::setDesktopSize(int x, int y, int width, int height)
 {
     if (m_primaryScreen)
-        QMetaObject::invokeMethod(m_primaryScreen, "setAvailableGeometry", Qt::AutoConnection, Q_ARG(QRect, QRect(0,0,width, height)));
+        QMetaObject::invokeMethod(m_primaryScreen, "setAvailableGeometry", Qt::AutoConnection, Q_ARG(QRect, QRect(x,y,width, height)));
 }
 
 void QAndroidPlatformIntegration::setDisplayMetrics(int width, int height)
--- a/src/plugins/platforms/android/qandroidplatformintegration.h
+++ b/src/plugins/platforms/android/qandroidplatformintegration.h
@@ -93,7 +93,7 @@ public:
     QAndroidPlatformScreen *screen() { return m_primaryScreen; }
     QPlatformOffscreenSurface *createPlatformOffscreenSurface(QOffscreenSurface *surface) const override;
 
-    virtual void setDesktopSize(int width, int height);
+    virtual void setDesktopSize(int x, int y, int width, int height);
     virtual void setDisplayMetrics(int width, int height);
     void setScreenSize(int width, int height);
     bool isVirtualDesktop() { return true; }
--- a/src/plugins/platforms/android/qandroidplatformscreen.cpp
+++ b/src/plugins/platforms/android/qandroidplatformscreen.cpp
@@ -323,8 +323,14 @@ void QAndroidPlatformScreen::setPhysicalSize(const QSize &size)
 
 void QAndroidPlatformScreen::setSize(const QSize &size)
 {
-    m_size = size;
-    QWindowSystemInterface::handleScreenGeometryChange(QPlatformScreen::screen(), geometry(), availableGeometry());
+    if (size != m_size) {
+        m_size = size;
+        QWindowSystemInterface::handleScreenGeometryChange(QPlatformScreen::screen(), geometry(), availableGeometry());
+        if (m_id != -1) {
+            releaseSurface();
+            QtAndroid::setSurfaceGeometry(m_id, geometry());
+        }
+    }
 }
 
 void QAndroidPlatformScreen::setAvailableGeometry(const QRect &rect)
@@ -350,11 +356,6 @@ void QAndroidPlatformScreen::setAvailableGeometry(const QRect &rect)
             }
         }
     }
-
-    if (m_id != -1) {
-        releaseSurface();
-        QtAndroid::setSurfaceGeometry(m_id, rect);
-    }
 }
 
 void QAndroidPlatformScreen::applicationStateChanged(Qt::ApplicationState state)
@@ -418,7 +419,7 @@ void QAndroidPlatformScreen::doRedraw(QImage* screenGrabImage)
     }
     QMutexLocker lock(&m_surfaceMutex);
     if (m_id == -1 && m_rasterSurfaces) {
-        m_id = QtAndroid::createSurface(this, m_availableGeometry, true, m_depth);
+        m_id = QtAndroid::createSurface(this, geometry(), true, m_depth);
         AndroidDeadlockProtector protector;
         if (!protector.acquire())
             return;
